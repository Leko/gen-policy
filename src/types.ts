export type DependencyRedirection = Record<string, string | boolean>

export type PolicyItem = {
  integrity: string
  dependencies?: DependencyRedirection
}

export type Policy = {
  onerror: "exit" | "log" | "throw"
  resources: Record<string, PolicyItem>
}
