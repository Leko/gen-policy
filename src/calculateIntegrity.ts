import crypto from "crypto"

export function calculateIntegrity(content: string): string {
  return (
    "sha384-" +
    crypto
      .createHash("sha384")
      .update(content)
      .digest("base64")
  )
}
