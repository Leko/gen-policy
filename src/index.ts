export { calculateIntegrity } from "./calculateIntegrity"
export { generatePolicies } from "./generatePolicies"
export * from "./types"
