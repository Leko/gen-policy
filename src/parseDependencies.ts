import path from "path"
import { parse } from "@babel/parser"
import traverse from "@babel/traverse"
import { sync as enhancedResolve } from "enhanced-resolve"
// @ts-ignore
import isBuiltinModule from "is-builtin-module"
import { DependencyRedirection } from "./types"

export function parseDependencies(
  code: string,
  filePath: string
): DependencyRedirection {
  const dependencies: DependencyRedirection = {}
  const ast = parse(code)

  traverse(ast, {
    CallExpression({ node }) {
      if (node.callee.type !== "Identifier" || node.callee.name !== "require") {
        return
      }
      if (node.arguments[0].type !== "StringLiteral") {
        // TODO: warning log
        return
      }

      const source = node.arguments[0].value
      if (isBuiltinModule(source)) {
        dependencies[source] = true
      } else {
        dependencies[source] = enhancedResolve(path.dirname(filePath), source)
      }
    },
  })

  return dependencies
}
