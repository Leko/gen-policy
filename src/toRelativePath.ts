import path from "path"

export function toRelativePath(from: string, to: string): string {
  const relativePath = path.relative(from, to)
  if (!relativePath.startsWith(`..${path.sep}`)) {
    return `.${path.sep}${relativePath}`
  }

  return relativePath
}
