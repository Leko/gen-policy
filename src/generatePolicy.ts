import { promises as fs } from "fs"
import { PolicyItem, DependencyRedirection } from "./types"
import { calculateIntegrity } from "./calculateIntegrity"
import { parseDependencies } from "./parseDependencies"
import { toRelativePath } from "./toRelativePath"

export async function generatePolicy(
  entry: string,
  manifestDir: string
): Promise<Record<string, PolicyItem>> {
  const relativePath = toRelativePath(manifestDir, entry)
  const content = await fs.readFile(entry, "utf8")
  const integrity = calculateIntegrity(content)

  if (entry.endsWith(".json")) {
    return {
      [relativePath]: {
        integrity,
      },
    }
  }
  const rawDependencies = parseDependencies(content, entry)

  const dependencies: DependencyRedirection = {}
  for (let source in rawDependencies) {
    const modulePath = rawDependencies[source]
    if (typeof modulePath === "boolean") {
      dependencies[source] = modulePath
    } else {
      dependencies[source] = toRelativePath(manifestDir, modulePath)
    }
  }

  const rest = await Promise.all(
    Object.values(rawDependencies)
      .filter((f): f is string => typeof f !== "boolean")
      .map(f => generatePolicy(f, manifestDir))
  )

  return {
    [relativePath]: {
      integrity,
      dependencies,
    },
    ...rest.reduce((acc, map) => ({ ...acc, ...map }), {}),
  }
}
