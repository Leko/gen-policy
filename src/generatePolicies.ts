import path from "path"
import findUp from "find-up"
import { Policy } from "./types"
import { generatePolicy } from "./generatePolicy"

export async function generatePolicies(
  entry: string,
  manifestDir: string
): Promise<Pick<Policy, "resources">> {
  const packageJsonPath = await findUp("package.json", {
    cwd: path.dirname(entry),
  })

  return {
    resources: {
      ...(await generatePolicy(entry, manifestDir)),
      ...(packageJsonPath
        ? await generatePolicy(packageJsonPath, manifestDir)
        : {}),
    },
  }
}
