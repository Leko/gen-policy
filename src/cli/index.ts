#!/usr/bin/env node
import fs from "fs"
import path from "path"
import chalk from "chalk"
import { prompt } from "enquirer"
import { options } from "./options"
import { generatePolicies } from "../generatePolicies"
import { calculateIntegrity } from "../calculateIntegrity"

async function parseArgs() {
  const cliOptions = options.parse()
  const entry = cliOptions["entry"] as string
  const outFile = cliOptions["out-file"]
  const outDir = cliOptions["out-dir"]
  const force = cliOptions["force"]

  if (!fs.existsSync(entry)) {
    throw new Error(`Entry file ${entry} doesn't exists`)
  }
  if (!fs.existsSync(outDir)) {
    throw new Error(`Entry file ${entry} doesn't exists`)
  }
  const out = path.join(outDir, outFile)
  if (fs.existsSync(out) && !force) {
    const { override } = await prompt({
      name: "override",
      type: "confirm",
      message: `${out} already exists. Are you sure you want to continue?`,
    })
    if (!override) {
      console.log("Canceled")
    }
  }
  return {
    entry,
    outFile,
    outDir,
  }
}

parseArgs()
  .then(({ entry, outFile, outDir }) => {
    const out = path.join(outDir, outFile)
    const relativeOut = path.relative(process.cwd(), out)
    return generatePolicies(entry, outDir).then(policy => {
      const content = JSON.stringify(policy, null, 2)
      fs.writeFileSync(out, content)
      const integrity = calculateIntegrity(content)
      const cmd = chalk.bold.cyan(
        `node --experimental-policy=${relativeOut} --policy-integrity=${integrity} ${entry}`
      )
      console.log(chalk.dim(`The policy file was saved to ${relativeOut}.`))
      console.log(
        chalk.dim(
          `To run Node.js with this policy, run the following command:\n`
        )
      )
      console.log("  $ " + cmd)
    })
  })
  .catch(e => {
    console.error(e.stack)
    process.exit(1)
  })
