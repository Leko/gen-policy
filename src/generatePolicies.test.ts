import fs from "fs"
import path from "path"
import { generatePolicies } from "../src"

const BASE_DIR = path.join(__dirname, "..", "fixtures")

describe("E2E", () => {
  const dirNames = fs
    .readdirSync(BASE_DIR)
    .map(dirName => path.join(BASE_DIR, dirName))
    .filter(dir => fs.statSync(dir).isDirectory())
  const table = dirNames.map(dir => [path.relative(process.cwd(), dir)])

  test.each(table)("Snapshot: %s", async dir => {
    const entry = path.join(dir, "entry.js")
    const manifestDir = path.join(dir, "policy.json")
    expect(await generatePolicies(entry, manifestDir)).toMatchSnapshot()
  })
})
