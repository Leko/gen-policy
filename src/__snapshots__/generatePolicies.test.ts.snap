// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`E2E Snapshot: fixtures/simple 1`] = `
Object {
  "resources": Object {
    "../../../package.json": Object {
      "integrity": "sha384-caD3G/oidzeDBY6lgIcKZcjvTywDHi+LPGICbkMwiJVmtddYYWrCMV4CvkomWDw/",
    },
    "../entry.js": Object {
      "dependencies": Object {
        "./foo.js": "../foo.js",
      },
      "integrity": "sha384-W1P8/wqNVrr0aAOTqbvFaQnbc6EZXHh8Cp1Id0YNOFmG9xVWmovdlBm7BB2vVFgW",
    },
    "../foo.js": Object {
      "dependencies": Object {},
      "integrity": "sha384-TA6OPoLSABbGGyrDx1Al7pxHP8hQF8Xlicz4+ZhTQ1oZQFdRtkKPLmMovCgztftC",
    },
  },
}
`;

exports[`E2E Snapshot: fixtures/with-builtin-module 1`] = `
Object {
  "resources": Object {
    "../../../package.json": Object {
      "integrity": "sha384-caD3G/oidzeDBY6lgIcKZcjvTywDHi+LPGICbkMwiJVmtddYYWrCMV4CvkomWDw/",
    },
    "../entry.js": Object {
      "dependencies": Object {
        "crypto": true,
        "fs": true,
        "path": true,
        "zlib": true,
      },
      "integrity": "sha384-bGSMKUIryN9WBtzKCdh/lxAazUSDPMpI++2psUBviYxA09GxK6EKF8E//PGH91gy",
    },
  },
}
`;

exports[`E2E Snapshot: fixtures/with-nested-require 1`] = `
Object {
  "resources": Object {
    "../../../package.json": Object {
      "integrity": "sha384-caD3G/oidzeDBY6lgIcKZcjvTywDHi+LPGICbkMwiJVmtddYYWrCMV4CvkomWDw/",
    },
    "../a.js": Object {
      "dependencies": Object {
        "./b": "../b.js",
      },
      "integrity": "sha384-27hwvFL0g1cSlpwuzwkapXAw1ojKWvWwMRnCSSfTdkL0OkUoZivotXG6pX76xh6W",
    },
    "../b.js": Object {
      "dependencies": Object {
        "./c": "../c.js",
      },
      "integrity": "sha384-uAgDhBvyptalrOJtz8jHcciHhGU90lhLUEB3efVmkfCby9jenrjZ86QR+w/tSIcE",
    },
    "../c.js": Object {
      "dependencies": Object {},
      "integrity": "sha384-TA6OPoLSABbGGyrDx1Al7pxHP8hQF8Xlicz4+ZhTQ1oZQFdRtkKPLmMovCgztftC",
    },
    "../entry.js": Object {
      "dependencies": Object {
        "./a": "../a.js",
      },
      "integrity": "sha384-GRgy8p/cvLQPQE1pDAU+eTrxbZA6FOqsEzotrduiIkBMXuktUtJ61+mR2A6YW/Za",
    },
  },
}
`;
